### PR TITLE
[IR][Optimization] Generalize optimization of Insert/Extract to TensorView.

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
Make the IR optimization of specializing the Insert/Extract instructions to TensorView more generic.
Added a more generic utility function called `isSliceMemoryContiguous`.

**Test Plan**
Some problems arise in some existing unit tests involving the Slice operator for the OpenCL backend which do not work after this optimization. Same unit tests work for the other backends so there is something wrong with the OpenCL implementation.
Will add more specialized unit tests after the OpenCL problem is solved.

### CODE REVIEW

**Comment by mciprian13** [comment]
@jfix71 I created a separate PR for this. Can you point someone which has expertise in OpenCL so it can help me?
Thanks!

**Comment by jfix71** [comment]
Hi @mciprian13 -- please see https://github.com/pytorch/glow/pull/4500. It should fix the problem, so once that has landed and you rebase on top tests should pass and we can land this.

**Comment by mciprian13** [comment]
@jfix71 You were right! Tests are now passing. I did not add new unit tests but since the logic is only as good as the utility function `isSliceMemoryContiguous` I think maybe:
- I could put this function somewhere public like in Type.h since I think it could be reused in other places
- add a couple of unit tests for this function
WDYT?

**Comment by jfix71** [comment]
> I could put this function somewhere public like in Type.h since I think it could be reused in other places

SGTM

> add a couple of unit tests for this function

Fine to have tests for this function, but no matter what I think it'd be good to add to IROptTest to cover more cases, e.g. see the simple test `insertOptimizer` we have now.

**Comment by mciprian13** [comment]
@jfix71 I added some stronger unit tests for Insert/Extract optimizations in `IROptTest`. Have a look and if fine by you let's have this landed.
Thanks!

File: include/glow/Base/Tensor.h [hunk 1]
```diff
@@ -1045,6 +1045,53 @@ inline size_t getFlattenedOffset(llvm::ArrayRef<dim_t> strides,
   return index;
 }
 
+/// Helper function which \returns true if a slice with the shape \p sliceShape
+/// references fully a tensor with the shape \p tensorShape. This happens when
+/// the shapes are the same.
+inline bool isSliceFull(llvm::ArrayRef<dim_t> sliceShape,
```
Comment (jfix71) [inline include/glow/Base/Tensor.h:1051]: Isn't this the same as just checking if `sliceShape == tensorShape`? I don't think this is necessary to add, can just compare the two inputs directly there.

File: include/glow/Base/Tensor.h [hunk 2]
```diff
@@ -1045,6 +1045,53 @@ inline size_t getFlattenedOffset(llvm::ArrayRef<dim_t> strides,
...
+}
+
+/// Helper function which \returns true if a slice with the shape \p sliceShape
+/// referenced from a larger tensor with the shape \p tensorShape is contiguous
+/// in memory (assuming the tensor it is referenced from is contiguous). This
+/// happens when the slice dimensions:
+/// - Start with singleton dimensions (dimensions equal to 1).
+/// - Continue with a partially extracted dimension (one maximum).
+/// - End with fully extracted dimensions.
+inline bool isSliceContiguous(llvm::ArrayRef<dim_t> sliceShape,
```
Comment (jfix71) [inline include/glow/Base/Tensor.h:1070]: nit: Please move the body of this function into `Tensor.cpp`.

File: include/glow/Base/Tensor.h (hunk 1 above)
Comment (mciprian13) [reply include/glow/Base/Tensor.h]: Yes you`re right. I think I over-engineered this. I removed this and replaced with equality checks in the unit tests.

File: include/glow/Base/Tensor.h (hunk 2 above)
Comment (mciprian13) [reply include/glow/Base/Tensor.h]: Done.

**Comment by mciprian13** [comment]
@jfix71 Can we have this landed? Thanks!

**Pull request closed without merging**